### DISCUSSION TITLE
Unit Test Cross Compiling

### DISCUSSION THREAD

**Discussion by ewong-canadensys** [nasa/cFS#556] [Q&A] [answered]
**Version:**  IC: Caelum-rc4+dev19, draco-rc3

make test  & make lcov use .gcno files

**_Target system: ARM Cortex 8_**

The following the steps ran successfully:

```
sudo make distclean
make SIMULATION=native ENABLE_UNIT_TESTS=true prep
make
make install
```

Subfolders within build  contains *.gcno files

 **_Cross-compiling on an x86 Ubuntu system_**

_x86 system_

```
sudo make distclean
make SIMULATION=cpu1 ENABLE_UNIT_TESTS=true prep  // Where cross-compiling toolchains are set up
make
make install
```

**Subfolders within build did not contain *.gcno files**

_target system_

transfer the Makefile and build folders, rename build/cpu1 to build/native and execute

```
make test  <--  **runs but without gcno files**
make lcov
```

**Any suggestions on how to generate .gcno files when cross-compiling?**

**Comment by skliper** [comment, marked as answer]
The coverage flags get added here when not cross-compiling and on linux:
https://github.com/nasa/osal/blob/6e6afb4d3ea611f5d11e4e94a5adee595cc08b99/src/bsp/generic-linux/build_options.cmake#L20-L23

To add coverage for other cases it may just be easiest to add those flags from whatever toolchain.cmake file you are using.  for example in toolchain-cpu1.cmake just add:
```
set(UT_COVERAGE_COMPILE_FLAGS -pg --coverage)
set(UT_COVERAGE_LINK_FLAGS    -pg --coverage)
```

**Reply by ewong-canadensys** [reply]
Thanks for the response,

The whole build folder was moved to the target after cross-compiling and the build/cpu1 folder was renames the build/native 

1. I think some further investigation is needed. As noted in the link, the cross-compiled folder probably need to be the same path for cFS since the absolute path may be encoded in the *.gcno files, might try this next

```
d@bb:~/test/cfs_aquila$ make test
lcov --capture --initial --directory build/native/default_cpu1 --output-file build/native/default_cpu1/coverage_base.info
Capturing coverage data from build/native/default_cpu1
Found gcov version: 8.3.0
Scanning build/native/default_cpu1 for .gcno files ...
Found 124 graph files in build/native/default_cpu1
Processing time/ut-coverage/CMakeFiles/coverage-time-ALL-object.dir/__/fsw/src/cfe_time_utils.c.gcno
geninfo: WARNING: could not open /home/ca/lrm/cfs_aquila/cfe/modules/core_api/fsw/inc/cfe_sb.h
```

running lcov on the target  is shown below, as noted it doesn't contain the .gcda files
```
debian@beaglebone:~/test/cfs_aquila$ make lcov
lcov --capture --rc lcov_branch_coverage=1 --directory build/native/default_cpu1 --output-file build/native/default_cpu1/coverage_test.info
Capturing coverage data from build/native/default_cpu1
Found gcov version: 8.3.0
Scanning build/native/default_cpu1 for .gcda files ...
geninfo: WARNING: no .gcda files found in build/native/default_cpu1 - skipping!
Finished .info-file creation
lcov --rc lcov_branch_coverage=1 --add-tracefile build/native/default_cpu1/coverage_base.info --add-tracefile build/native/default_cpu1/coverage_test.info --output-file build/native/default_cpu1/coverage_total.info
Combining tracefiles.
Reading tracefile build/native/default_cpu1/coverage_base.info
Reading tracefile build/native/default_cpu1/coverage_test.info
lcov: ERROR: no valid records found in tracefile build/native/default_cpu1/coverage_test.info
```

**Reply by skliper** [reply]
Right, there's a slew of issues to address when actually trying to run coverage on target.  Last I looked there were some pretty good resources on line for the steps involved, but it's going to be somewhat target dependent.  cFS doesn't have an out-of-the-box solution at this point.

**Reply by ewong-canadensys** [reply]
Thanks for your help
